_posts/2022-05-18-My-OTP-25-highlights.md

```diff
@@ -0,0 +1,713 @@
+unsuccessful matches within the same construct without risking to confuse
+happy and unhappy paths.
+
+Given the structure described here, the final expression may look like:
+
+```erlang
+maybe
+    Foo = bar(),            % normal exprs still allowed
+    {ok, X} ?= f(Foo),
+    [H|T] ?= g([1,2,3]),
+    ...
+else
+    {error, Y} ->
+        {ok, "default"};
+    {ok, _Term} ->
+        {error, "unexpected wrapper"}
+end
+```
+
+### Motivation
+
+With the `maybe` construct it is possible to reduce deeply nested conditional expressions and make messy patterns found in the wild unnecessary. It also provides a better separation of concerns when implementing functions.
+
+#### Reducing Nesting
+
+One common pattern that can be seen in Erlang is deep nesting of `case
+... end` expressions, to check complex conditionals.
+
+Take the following code taken from
+[Mnesia](https://github.com/erlang/otp/blob/a0ae44f324576104760a63fe6cf63e0ca31756fc/lib/mnesia/src/mnesia_backup.erl#L106-L126),
+for example:
+
+```erlang
+commit_write(OpaqueData) ->
+    B = OpaqueData,
+    case disk_log:sync(B#backup.file_desc) of
+        ok ->
+            case disk_log:close(B#backup.file_desc) of
+                ok ->
+                    case file:rename(B#backup.tmp_file, B#backup.file) of
+                        ok ->
+                            {ok, B#backup.file};
+                        {error, Reason} ->
+                            {error, Reason}
+                    end;
+                {error, Reason} ->
+                    {error, Reason}
+            end;
+        {error, Reason} ->
+            {error, Reason}
+    end.
+```
+
+The code is nested to the extent that shorter aliases must be introduced
+for variables (`OpaqueData` renamed to `B`), and half of the code just
+transparently returns the exact values each function was given.
+
+By comparison, the same code could be written as follows with the new
+construct:
+
+```erlang
+commit_write(OpaqueData) ->
+    maybe
+        ok ?= disk_log:sync(OpaqueData#backup.file_desc),
+        ok ?= disk_log:close(OpaqueData#backup.file_desc),
+        ok ?= file:rename(OpaqueData#backup.tmp_file, OpaqueData#backup.file),
+        {ok, OpaqueData#backup.file}
+    end.
+```
+
+Or, to protect against `disk_log` calls returning something else than `ok |
+{error, Reason}`, the following form could be used:
+
+```erlang
+commit_write(OpaqueData) ->
+    maybe
+        ok ?= disk_log:sync(OpaqueData#backup.file_desc),
+        ok ?= disk_log:close(OpaqueData#backup.file_desc),
+        ok ?= file:rename(OpaqueData#backup.tmp_file, OpaqueData#backup.file),
+        {ok, OpaqueData#backup.file}
+    else
+        {error, Reason} -> {error, Reason}
+    end.
+```
+
+The semantics of these calls are identical, except that it is now
+much easier to focus on the flow of individual operations and either
+success or error paths.
+
+# Compiler news
```

Comment:
The binary syntax things are described in detail later in the post. Not sure the -nifs attribute qualifies for a highlight.

Reply:
Agree on this too, removed

Reply:
agree, removed